### ISSUE TITLE
Please make right-click menu "Open with" available on macOS

### ISSUE THREAD

**Issue by Andreas0602** [mucommander/mucommander#759]
On macOS, please make available the right-click menu item "Open with" (like in Finder).

I'm aware that it's possible to customize commands.xml. That's better than nothing, but it's not a full replacement of the "real" "Open with" menu, particularly because you have to do it manually, while the "real" "Open with" menu is automatically populated by the system. I'm also aware that this might be difficult to implement in Java, but on the other hand I believe it's worth to spend some effort for this, because it's an important feature for a file manager.

**Comment by ahadas** [comment]
> I'm also aware that this might be difficult to implement in Java

that's quite an understatement :)
if one is able to do it, I'd say it would be better to implement it as a library that can also be used by other applications.
it's out of the scope of mucommander to implement something like that but if there's such a library, please let us know and we'll try to integrate it

**Comment by Andreas0602** [comment]
I'm shocked to hear that this is "out of scope" of muCommander. I'm not aware of any other file manager that does **not** have this feature. As a user, I don't care whether it's implemented in a library or directly in muCommander, but it's a feature that's definitely needed, better sooner than later. Sorry, but that's my point of view.

**Comment by ahadas** [comment]
I meant that implementing this functionality is out of scope. unlike many other file managers, muCommander is a cross-platform application, we cannot afford implementing operating system-specific extensions like that - it would take significant portion of our time and doesn't align with the unified experience we aim for.

I understand that as a user you don't care how the feature is implemented (as a library or within muCommander) because you're interested in the functionality itself - I shared that view for others that may see this issue in the future and would like to help with that

**Comment by Andreas0602** [comment]
I don't think that this feature would break the unified experience. Actually, from a user's point of view, it would look and feel the same way on all platforms. Double Commander is also a cross-platform application, and it does have this feature. But of course it's your decision which features you invest your time in.

By the way, I don't think that many other users will look into this issue and then help with it, because you have closed this issue. "Closed" normally means: The issue is solved/implemented, or it's impossible to solve/implement it. I really hope that Java is not so limited in its capabilities that it's impossible to implement this feature. Otherwise, Java would be a bad choice for this project.

**Comment by ahadas** [comment]
ok, let's reopen this and add label it properly

**Comment by pskowronek** [comment]
Sometime ago I've also investigated that how to get the list of apps that can open a given file type - not so easy on macOS:
https://apple.stackexchange.com/questions/342229/anyway-to-obtain-list-of-application-that-can-open-a-file-via-terminal

For windows:
https://stackoverflow.com/questions/65569438/macos-how-to-open-the-open-with-menue-from-code

(Just to keep the useful links handy)

**Comment by pskowronek** [comment]
On stackexchange post from above somebody has mentioned [Duti](http://duti.org/) tool (written in C). I've checked it - it is open-source and available on github, it is still maintained and it works. It returns the list of app bundle ids for a given file or extension, then using `open -b` we can open such app by bundle id. What's cool about Duti is that it is available via `brew` - so my idea would be to check if `duti` is present in the system if so, then we could produce the list for `Open with`.  To inform the user that they need to install `duti` we could add a new option to enable this feature in Preferences and there a comment could appear to run `brew install duti` if `duti` is not present as a command in the system.

@ahadas, @Andreas0602 What do you think?

**Comment by ahadas** [comment]
Putting #898 aside (though I'll say that I truly hope that issue will be fixed as it might affect users of muCommander in several countries in some flows), I think the proposed solution for open-with using native tools on macOS is an interesting direction. We also rely on a non-cross platform command line tool (skopeo) for inspecting container registries and we first check its present, and if it's not available the 'registry' protocol is disabled - I'm noting this as we had an issue identifying whether the tool exists, it sounds trivial but we had several attempts until we found [the right way ](https://github.com/mucommander/mucommander/commit/4f452e2d9d789cad32b35a04edbb53d3daf3e307). I can imagine that the os-api will allow us to use `duti`, if exists, on macOS only - I think it is worth a try

**Comment by ahadas** [comment]
> > I truly hope that issue will be fixed
> 
> Hoping for a fix from Apple is good, fixing it yourself in muCommander would be better. wink

@Andreas0602 we again diverge from the topic that should be discussed here - we can continue this in #898, but in short - I would have agreed we should fix it in muCommander if it was an issue on our side or if we had a simple workaround for it. I'm afraid that in order to avoid this we would need to rename/copy files using Apple scripts, which would break our abstraction of using streams from any format/protocol to any format/protocol. as it happens only in some rare flows after the fix we merged in 1.2.0-2 and it's an issue in the operating system that affects other applications as well - I don't think it would be wise to make such major changes in muCommander. if you see a better way to fix it in muCommander, please suggest it in #898

**Comment by pskowronek** [comment]
PR provided, still WiP - awaiting impl comments.

**Comment by pskowronek** [comment]
Can you check logs?

**Comment by pskowronek** [comment]
A stacktrace should be logged and could be more descriptive, but I guess it is turned off?

How did you install `duti`? Via brew? Can you check where `duti` is installed in the system? (what is its path).
Have you restarted muC?

**Comment by pskowronek** [comment]
hmmm, you had `brew` already installed before running muC, right? Do you have `/usr/local/git/bin/` in `PATH`?

**Comment by pskowronek** [comment]
Sorry I meant `/usr/local/bin/`

**Comment by pskowronek** [comment]
> However, when I build and run muCommander via `./gradlew clean run`, it does work!
Cool!

Do you have PATH configured in ~/.profile or only in ~/.zshrc / ~/.zshenv / ~/.zprofile?

As @ahadas has pointed out "it sounds trivial but we had several attempts until @danielerez found [the right way ](https://github.com/mucommander/mucommander/commit/4f452e2d9d789cad32b35a04edbb53d3daf3e307)." - but I would like to avoid having hard coded paths (differences where macports, brew, fink keep their bins), plus macbooks with arm - I guess `brew` keeps them in another path.

> By the way, some icons are broken. Besides that, it works great!

yeah, I know, haven't checked it yet, this is minor thing at this point.

**Comment by pskowronek** [comment]
How do you set PATH then?

**Comment by pskowronek** [comment]
Sorry to bother, probably you need to restart mac to feed PATH from ~/.profile

**Comment by pskowronek** [comment]
> > Homebrew page suggests [this](https://docs.brew.sh/FAQ#my-mac-apps-dont-find-homebrew-utilities). @Andreas0602 can you give it a try? For me it works now.
> 
> For me, this works too. So at least we have one solution now that does work. But it's a bit inconvenient: Messing around with sudo, rebooting the machine etc. Also, I would be interested in how to undo the modification made by the sudo command.

Agreed, not exactly user friendly.

> How about letting muCommander start a terminal, run the command `which duti`, and capture the output?

This idea sounds promising.

> Also, `` echo `brew --prefix`/bin `` will give you the path.

Somebody may use macports or fink instead....

**Comment by pskowronek** [comment]
@Andreas0602 pushed a change that uses `bash -l` and `which duti` (great suggestion btw).
To revert that sudo operation run `sudo launchctl config user path /usr/bin:/bin:/usr/sbin:/sbin` (this is a default on mac), then reboot and retry with dmg -> .app.

**Comment by pskowronek** [comment]
> Maybe I broke something in my environment when messing around with the PATH.

Yup :/

**Comment by pskowronek** [comment]
PR updated with Preferences.

> By the way, some icons are broken.  
BTW, @Andreas0602 - apparently it has mysteriously healed itself :) I cannot reproduce this anymore...

**Comment by pskowronek** [comment]
When you open terminal (F12) then you have zsh? Is duti working over there?

**Comment by pskowronek** [comment]
hmmm, I've switched to `zsh` and it also works for me. 

Can you paste what you get in `Debug console`?

**Comment by pskowronek** [comment]
PR updated @Andreas0602

**Comment by pskowronek** [comment]
> May I ask why there is an option for that functionality? No other file manager has such an option. The functionality is always enabled.

Fair question, so my thinking was:

1.  it might be annoying to wait for those apps to appear - ppl may have their own custom apps defined and they want it to show instantly (ok, this is kinda already done, apps are being loaded in the background while custom apps show instantaneously)
2.  since it doesn't work out-of-the-box (it requires `duti`) I wanted to have place where somebody will be hinted how to make it work (see PR with screenshot, an error dialog box /maybe it should be 'info dialog box'?) will appear when selecting that feature and `duti` is not present).

**Comment by pskowronek** [comment]
Interesting idea. @ahadas what's your opinion?

**Comment by ahadas** [comment]
there is a dialog with title "Welcome to muCommander" that is shown for the first time the application starts (by checking if the a preferences file exists so you can get this dialog by removing the preferences folder) - when starting on macOS and can't find 'duti', we can add a message there. it sounds good

**Comment by pskowronek** [comment]
but if somebody just updates muC this dialog won't show, right?

**Comment by ahadas** [comment]
yeah, that's right. so how about using the 'snapshot' configuration for this?
I still think that when installing muCommander from scratch it makes sense to display a message about 'duti' not being available on that welcome screen. and if that welcome screen is not shown, then we can check the 'snapshot' configuration

**Comment by pskowronek** [comment]
@ahadas  "then we can check the 'snapshot' configuration" - can you explain more? I don't get this 'snapshot' configuration idea (and what is snapshot configuration at all....)

**Comment by ahadas** [comment]
@pskowronek sure, there are two kinds of configuration: (1) that gets default values but can change through the preferences dialog - that group is stored in the preferences.xml file; and (2) that stores the last state the application has in its previous run, e.g., the amount of open tabs and their locations, that is stored in the snapshot.xml file. the snapshot configuration refers to the second type.
what I have in mind is to say that if the snapshot.xml file doesn't exist or doesn't have the property for 'duti' then we consider this as if 'duti' should be used (and then if the preferences.xml file doesn't exist, we can have the message in the welcome screen, and otherwise in a different dialog). otherwise, the snapshot.xml configuration would include a property that indicates whether 'duti' was used - if it's set to false, it means that the user has already been notified about that and therefore nothing should be done if 'duti' is not detected. if it's set to true and 'duti' is not detected, we can present a message to the user. what do you think?

**Comment by pskowronek** [comment]
"we can present a message to the user. what do you think?" - the message like ErrorDialog (see my PR) or via "Welcome to muCommander"?

**Comment by ahadas** [comment]
@pskowronek you mean when upgrading right? if so, then yeah, in that case a dialog (not necessarily an ErrorDialog, it's more an informative dialog - some users would want to install 'duti' and some probably wouldn't) makes sense

**Comment by pskowronek** [comment]
So, @ahadas & @Andreas0602  your idea were nice, but I decided to propose a 3rd option. Since I don't like pop-up dialogs appearing unexpectedly, I've simply added a menu item 'Enable apps here?' that if clicked will show info dialog how to show apps in that menu. See updated #917 screenshots over there.
